src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorFileChangeDetector.cs

```diff
@@ -0,0 +1,151 @@
+        {
+            if (workspaceDirectory is null)
+            {
+                throw new ArgumentNullException(nameof(workspaceDirectory));
+            }
+
+            // Dive through existing Razor files and fabricate "added" events so listeners can accurately listen to state changes for them.
+
+            workspaceDirectory = _filePathNormalizer.NormalizeForRead(workspaceDirectory);
+            var existingRazorFiles = GetExistingRazorFiles(workspaceDirectory);
+
+            await Task.Factory.StartNew(() =>
+            {
+                foreach (var razorFilePath in existingRazorFiles)
+                {
+                    FileSystemWatcher_RazorFileEvent(razorFilePath, RazorFileChangeKind.Added);
+                }
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+
+            // This is an entry point for testing
+            OnInitializationFinished();
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                // Client cancelled connection, no need to setup any file watchers. Server is about to tear down.
+                return;
+            }
+
+            // Start listening for project file changes (added/removed/renamed).
+
+            for (var i = 0; i < RazorFileExtensions.Count; i++)
+            {
+                var extension = RazorFileExtensions[i];
+                var watcher = new FileSystemWatcher(workspaceDirectory, "*" + extension)
+                {
+                    NotifyFilter = NotifyFilters.FileName | NotifyFilters.LastWrite,
+                    IncludeSubdirectories = true,
+                };
+
+                watcher.Created += (sender, args) => FileSystemWatcher_RazorFileEvent_Background(args.FullPath, RazorFileChangeKind.Added);
+                watcher.Deleted += (sender, args) => FileSystemWatcher_RazorFileEvent_Background(args.FullPath, RazorFileChangeKind.Removed);
+                watcher.Renamed += (sender, args) =>
+                {
+                    // Translate file renames into remove->add
+
+                    if (args.OldFullPath.EndsWith(extension, FilePathComparison.Instance))
+                    {
+                        // Renaming from Razor file to something else.
+                        FileSystemWatcher_RazorFileEvent_Background(args.OldFullPath, RazorFileChangeKind.Removed);
+                    }
+
+                    if (args.FullPath.EndsWith(extension, FilePathComparison.Instance))
+                    {
+                        // Renaming to a Razor file.
+                        FileSystemWatcher_RazorFileEvent_Background(args.FullPath, RazorFileChangeKind.Added);
+                    }
+                };
+
+                _watchers.Add(watcher);
+            }
+
+            for (var i = 0; i < _watchers.Count; i++)
+            {
+                _watchers[i].EnableRaisingEvents = true;
```

Comment:
Is there any reason why this has to happen after both watchers have been created and added as opposed to within the previous for? Also, why `for (var i =0` instead of `foreach(var x in y)` if we're not using `i` for anything but the index of this List?

Reply:
If there IS a specific reason why these patterns are needed we should probably leave a comment describing it so somebody (like me) doesn't come through trying to "clean it".

Reply:
Hmm, we could probably include it in the above for statement, it just means that potential files could trigger the file watcher before both are setup. That's technically true here as well anyhow because of the race between the first setting of `EnableRaisingEvents` and the second.

As for why `for (var i = 0` vs. `foreach (var x in y)`. Foreach allocates an enumerator where the prior does not. If you have a functional equivalence between the two without significant code change then it's good practice to prefer the non-foreach variant.